lib/reek/smells/smell_repository.rb

```diff
@@ -23,37 +25,32 @@ def initialize(smell_types: self.class.smell_types,
                      configuration: {})
         @configuration = configuration
         @smell_types   = smell_types
+        @detectors     = smell_types.map do |klass|
+          { klass => klass.new(configuration_for(klass)) }
+        end.reduce({}, :merge)
       end
 
-      def report_on(listener)
-        detectors.each_value { |detector| detector.report_on(listener) }
+      def report_on(collector)
+        detectors.each_value { |detector| detector.report_on(collector) }
       end
 
       def examine(context)
-        smell_listeners[context.type].each do |detector|
-          detector.examine(context)
+        smell_detectors_for(context.type).each do |detector|
+          detector.run_for(context)
         end
       end
 
-      def detectors
-        @initialized_detectors ||= smell_types.map do |klass|
-          { klass => klass.new(source_configuration_for(klass)) }
-        end.reduce({}, :merge)
-      end
-
       private
 
-      private_attr_reader :configuration, :smell_types
-
-      def source_configuration_for(klass)
-        configuration[klass] || {}
+      def configuration_for(klass)
+        configuration.fetch klass, {}
       end
 
-      # TODO: Make a method smell_detectors_for(scope)
-      def smell_listeners
-        @smell_listeners ||= Hash.new { |hash, key| hash[key] = [] }.tap do |listeners|
-          detectors.each_value { |detector| detector.register(listeners) }
+      def smell_detectors_for(type)
+        @smell_detectors_for ||= Hash.new { |hash, key| hash[key] = [] }.tap do |hooks|
+          detectors.each_value { |detector| detector.register(hooks) }
```

Comment:
I find it confusing either way. What it *is* is a map from context to an array of smell detectors. Let's name it accordingly.

Also, we only ever use the values of detectors, so let's make it an Array instead of a Hash.

Finally, perhaps we can remove the indirection by moving the logic of SmellDetector#register her. It is not very complex.